File: java/src/museo/EncargadoCatalogo.java
package museo;

public class EncargadoCatalogo extends Usuario {

    private final String codigoEmpleado;
    private final CatalogoObras catalogo;

    public EncargadoCatalogo(
            int idUsuario,
            String nombre,
            String correo,
            boolean activo,
            String contrasena,
            String codigoEmpleado,
            CatalogoObras catalogo
    ) {
        super(idUsuario, nombre, correo, activo, contrasena);
        this.codigoEmpleado = codigoEmpleado;
        this.catalogo = catalogo;
    }

    public String getCodigoEmpleado() {
        return codigoEmpleado;
    }

    public CatalogoObras getCatalogo() {
        return catalogo;
    }

    public void registrarObra(ObraArte obra) {
        if (!autenticar()) {
            throw new IllegalStateException("Debe autenticarse antes de registrar obras.");
        }
        catalogo.agregarObra(obra);
    }

    public void actualizarObra(ObraArte obra) {
        if (!autenticar()) {
            throw new IllegalStateException("Debe autenticarse antes de actualizar obras.");
        }
        if (!catalogo.contieneObra(obra)) {
            catalogo.agregarObra(obra);
        }
        catalogo.tocar();
    }

    public void asignarSala(ObraArte obra, Sala sala) {
        if (!autenticar()) {
            throw new IllegalStateException("Debe autenticarse antes de asignar salas.");
        }
        sala.agregarObra(obra);
    }
}

File: javascript/src/museo/encargadoCatalogo.js
import { Usuario } from "./usuario.js";

export class EncargadoCatalogo extends Usuario {
  constructor(idUsuario, nombre, correo, activo, contrasena, codigoEmpleado, catalogo) {
    super(idUsuario, nombre, correo, activo, contrasena);
    this.codigoEmpleado = codigoEmpleado;
    this.catalogo = catalogo;
  }

  getCodigoEmpleado() {
    return this.codigoEmpleado;
  }

  getCatalogo() {
    return this.catalogo;
  }

  registrarObra(obra) {
    if (!this.autenticar()) {
      throw new Error("Debe autenticarse antes de registrar obras.");
    }
    this.catalogo.agregarObra(obra);
  }

  actualizarObra(obra) {
    if (!this.autenticar()) {
      throw new Error("Debe autenticarse antes de actualizar obras.");
    }
    if (!this.catalogo.contieneObra(obra)) {
      this.catalogo.agregarObra(obra);
    }
    this.catalogo.tocar();
  }

  asignarSala(obra, sala) {
    if (!this.autenticar()) {
      throw new Error("Debe autenticarse antes de asignar salas.");
    }
    sala.agregarObra(obra);
  }
}
